### DISCUSSION TITLE
Supress the creation of a Missed Image in -crop

### DISCUSSION THREAD

**Discussion by duarm** [ImageMagick/ImageMagick#7488] [Help] [answered]
### ImageMagick version

7.1.1-35

### OS
arch linux

### Description

I have a sheet like this:

```
XXXXXXXX
XXXXXXXX
XXXXXXXX
XXXXXXXX
XXXXXXXXXX
XXXXXXXXXX
XXXXXXXXXX
XXXXXXXXXX
```

cropping this sheet results in 8 missed images (transparent 1x1 output), as expected. Is there a way to supress the creation of these files, and if possible, ignore their index, outputting a continuous %d, instead of having gaps. E.g.

Problem
```sh
$ magick -crop 64x64 %d.png
$ ls 
1.png 2.png 3.png 4.png 5.png
# 2.png and 3.png are missed images (1x1 transparent)
 ```
 
Expected
```sh
$ magick -crop 64x64 %d.png
$ ls 
1.png 2.png 3.png
# 2.png and 3.png were ignored, now 4.png and 5.png were shifted back
 ```

**Comment by snibgo** [comment]
> $ magick -crop 64x64 %d.png

That isn't a valid command. It needs an input image.

> I have a sheet like this:

Perhaps each "x" or space represents a block of 64x64 pixels. So that image is 640 pixels wide and 512 pixels high. Is that correct? If so, then `-crop 64x64` should give 10*8 = 80 output images.

I probably misunderstand the problem. Please show a sample input, and the actual command.

**Reply by duarm** [reply]
Yes, everything you said is correct. Made a sample sheet to illustrate the problem
![char_a_p1_0bas_humn_v01](https://github.com/user-attachments/assets/89161f0d-5284-415a-89e3-cafdaa227529)

```sh
$ magick char_a_p1_0bas_humn_v01.png -crop 64x64 %d.png
```

char_a_p1_0bas_humn_v01.png being the embedded image, replicates my exact setup. IT creates some missed images because of that empty section, I want to ignore those sections instead of outputting an empty 14.png for example.

**Comment by fmw42** [comment]
What OS?

**Reply by duarm** [reply]
archlinux

**Comment by snibgo** [comment, marked as answer]
I can't see an easy way to do this. It could be done in a script. For example:
```
magick spriteBaby.png -crop 64x64 +repage -format %[fx:maxima.a==0?t:-1], info:
```
... makes this output:
```
-1,-1,-1,-1,-1,-1,6,7,-1,-1,-1,-1,-1,-1,14,15,-1,-1,-1,-1,-1,-1,22,23,-1,-1,-1,-1,-1,-1,30,31,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
```
So it is a list of which outputs we don't want (6, 7, 14, 15, etc) interspersed with "-1,". Use OS utilities to strip out each `-1,` and use another `magick` command, to include an option like:
```
-delete 6,7,14,15
```

**Reply by duarm** [reply]
thanks, think this is the best we got

**Comment by GeeMack** [comment]
Even when an ImageMagick command creates multiple output images, the command can test for conditions and name output files differently according to the results. This trick can be used to sort the outputs of a single command into groups or categories. 

Using your example sprite sheet of 64 images where 8 of them are blank, this command crops the input into 64 units, tests each unit for the value of "%[fx:mean]" to determine fully transparent or not, and sets the output filenames accordingly. The command...

    magick sprites.png -crop 64x64 -set filename:0 "%[fx:s.mean?0:9]" sprite_%[filename:0]-%02d.png

... will result in 64 output images. 56 of them will be named "sprite_0-NN.png" and contain an image. The other 8 of them will be named "sprite_9-NN.png", and these will be the empty, transparent images.

The "NN" part of the filenames will be the number each image had in the list *before* they get sorted by the "0" and "9". If having the list in strict order is important, run one more IM command to read in only the "sprite_0-*.png" set and output those as PNGs in sequence...

    magick  sprite_0-*.png finalsprite-%02d.png

Then go back and delete the intermediate images manually or with a short script.

That's about the simplest method I can come up with for ImageMagick to sort the "keepers" from the "discards", only it doesn't literally omit the discard images from the output.

**Reply by duarm** [reply]
Seems like an easier way to delete the blank images, unfortunate we can't tell imagemagick to not output those, thanks
